adapters/repos/db/lsmkv/segment_bloom_filters.go

```diff
@@ -0,0 +1,245 @@
+	}
+
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("write bloom filter to disk: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close bloom filter file: %w", err)
+	}
+
+	return nil
+}
+
+func (ind *segment) loadBloomFilterFromDisk() error {
+	f, err := os.Open(ind.bloomFilterPath())
+	if err != nil {
+		return fmt.Errorf("open file for reading: %w", err)
+	}
+
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(f); err != nil {
+		return fmt.Errorf("read bloom filter data from file: %w", err)
+	}
+	data := buf.Bytes()
+
+	chcksm := binary.LittleEndian.Uint32(data[:4])
+	actual := crc32.ChecksumIEEE(data[4:])
+	if chcksm != actual {
+		return ErrInvalidChecksum
+	}
+
+	ind.bloomFilter = new(bloom.BloomFilter)
+	_, err = ind.bloomFilter.ReadFrom(bytes.NewReader(data[4:]))
+	if err != nil {
+		return fmt.Errorf("read bloom filter from disk: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close bloom filter file: %w", err)
+	}
+
+	return nil
+}
+
+func (ind *segment) initSecondaryBloomFilter(pos int) error {
+	before := time.Now()
+
+	ok, err := fileExists(ind.bloomFilterSecondaryPath(pos))
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		err = ind.loadBloomFilterSecondaryFromDisk(pos)
+		if err == nil {
+			return nil
+		}
+
+		if err != ErrInvalidChecksum {
+			// not a recoverable error
+			return err
+		}
+
+		// now continue re-calculating
+	}
+
+	keys, err := ind.secondaryIndices[pos].AllKeys()
+	if err != nil {
+		return err
+	}
+
+	ind.secondaryBloomFilters[pos] = bloom.NewWithEstimates(uint(len(keys)), 0.001)
+	for _, key := range keys {
+		ind.secondaryBloomFilters[pos].Add(key)
+	}
+
+	if err := ind.storeBloomFilterSecondaryOnDisk(pos); err != nil {
+		return fmt.Errorf("store secondary bloom filter on disk: %w", err)
+	}
+	took := time.Since(before)
+
+	ind.logger.WithField("action", "lsm_init_disk_segment_build_bloom_filter_secondary").
+		WithField("secondary_index_position", pos).
+		WithField("path", ind.path).
+		WithField("took", took).
+		Debugf("building bloom filter took %s\n", took)
+	return nil
+}
+
+func (ind *segment) storeBloomFilterSecondaryOnDisk(pos int) error {
+	buf := new(bytes.Buffer)
+	_, err := ind.secondaryBloomFilters[pos].WriteTo(buf)
+	if err != nil {
+		return fmt.Errorf("write bloom filter: %w", err)
+	}
+
+	data := buf.Bytes()
+	chksm := crc32.ChecksumIEEE(data)
+
+	f, err := os.Create(ind.bloomFilterSecondaryPath(pos))
+	if err != nil {
+		return fmt.Errorf("open file for writing: %w", err)
+	}
+
+	if err := binary.Write(f, binary.LittleEndian, chksm); err != nil {
+		return fmt.Errorf("write checkusm to file: %w", err)
+	}
+
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("write bloom filter to disk: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close bloom filter file: %w", err)
+	}
+
+	return nil
+}
+
+func (ind *segment) loadBloomFilterSecondaryFromDisk(pos int) error {
```

Comment:
The handling for primary and secondary seems to be almost identical, can you move the identical parts into extra methods?

Reply:
Good point, let me check if there's a clean way of doing this.

Reply:
I didn't combine the primary/secondary, but since I refactored the storeWithChecksum/loadWithChecksum parts, where I've already unified the common logic (your other comment) the remaining methods are now super short :-)